File: java/src/main/java/dsa/blind150/binarySearch/BinarySearch.java
package dsa.blind150.binarySearch;

import dsa.utils.LoggerUtil;
import org.slf4j.Logger;

public class BinarySearch {
    private static final Logger logger = LoggerUtil.getLogger();

    public static int binarySearch(int[] arr, int target) {
        logger.info("binarySearch invoked with target: {}", target);
        int low = 0, high = arr.length - 1;

        while (low <= high) {
            int mid = low + ((high - low) / 2);
            logger.debug("Low: {}, High: {}, Mid: {}, Mid Value: {}", low, high, mid, arr[mid]);

            if (arr[mid] == target) {
                logger.info("Target {} found at index: {}", target, mid);
                return mid;
            } else if (arr[mid] < target) {
                logger.debug("Target {} is greater than {}, moving to the right half.", target, arr[mid]);
                low = mid + 1;
            } else {
                logger.debug("Target {} is less than {}, moving to the left half.", target, arr[mid]);
                high = mid - 1;
            }
        }
        logger.info("Target {} not found in the array.", target);
        return -1;
    }

    public static void main(String[] args) {
        int[] arr = {1, 2, 3, 4, 5, 6, 7, 8, 9};
        logger.info("Starting binary search tests.");

        // Test cases with expected results for clarity
        int[] testTargets = {5, 10, 1};
        String[] testDescriptions = {
                "Searching for target in the middle of the array",
                "Searching for a target not present in the array",
                "Searching for the smallest element"
        };

        // Execute test cases
        int testCaseNumber = 1;
        for (int target : testTargets) {
            logger.info("Test Case {}: {}", testCaseNumber, testDescriptions[testCaseNumber - 1]);
            int result = binarySearch(arr, target);
            logger.info("Test Case {} Result: Target {} found at index: {}", testCaseNumber, target, result);
            testCaseNumber++;
        }
    }
}

File: python/src/blind150/binarySearch/BinarySearch.py
from typing import List
from src.utils.logging_util import get_logger

logger = get_logger(__name__)


class BinarySearch:
    def searchTarget(self, target: int, nums: List[int]) -> int:
        logger.info("searchTarget method invoked with target: %d", target)
        left, right = 0, len(nums) - 1

        while left <= right:
            mid = left + ((right - left) // 2)
            logger.debug("Left: %d, Right: %d, Mid: %d, Mid Value: %d", left, right, mid, nums[mid])

            if nums[mid] == target:
                logger.info("Target %d found at index: %d", target, mid)
                return mid
            elif nums[mid] < target:
                logger.debug("Target %d is greater than %d, moving to the right half.", target, nums[mid])
                left = mid + 1
            else:
                logger.debug("Target %d is less than %d, moving to the left half.", target, nums[mid])
                right = mid - 1

        logger.info("Target %d not found in the array.", target)
        return -1


# Driver code to test the BinarySearch class and searchTarget method
if __name__ == "__main__":
    # Test cases
    nums = [1, 2, 3, 4, 5, 6, 7, 8, 9]

    test_cases = [5, 10, 1]
    for idx, target in enumerate(test_cases, start=1):
        logger.info("Test Case %d: Searching for target %d", idx, target)
        binary_search = BinarySearch()
        result = binary_search.searchTarget(target, nums)
        logger.info("Test Case %d Result: Target %d found at index: %d", idx, target, result)
